### PR TITLE
Integrate RNTester Redesign (8/26 Nightly Build)

### LINKED ISSUE #5775: E2ETest page list is hard-coded, make this data-driven
Currently the VisitAllPages test has a hard-coded list of RNTester pages:

```
let pages = [
  '<ActivityIndicator>',
  '<Button>',
  //  'Custom Views',
  '<DatePicker>',
  'Fast Path Texts',
  '<FlatList>',
  '<Flyout>',
  '<Glyph> UWP',
  '<Image>',
  //  '<FlatList> - MultiColumn',
  'New App Screen',
  '<Picker>',
  '<PickerWindows>',
  '<Pressable>',
  '<Popup>',
  'Keyboard extension Example',
  '<ScrollViewSimpleExample>',
  //  '<SectionList>',
  '<Switch>',
  '<Text>',
  '<TextInput>',
  //'<Touchable*> and onPress',
  '<TransferProperties>',
  '<TransparentHitTestExample>',
  '<View>',
  'Keyboard Focus Example',
  //  'Accessibility',
  'Accessibility Windows',
  'AsyncStorage Windows',
  'Alert',
  'Animated - Examples',
  'Animated - Gratuitous App',
  'Appearance',
  'AppState',
  'Border',
  'Box Shadow',
  'Clipboard',
  'Crash',
  'DevSettings',
  'Dimensions',
  'Keyboard',
  'Layout Events',
  'Linking',
  'Layout - Flexbox',
  'Mouse Events',
  'Native Animated Example',
  'PanResponder Sample',
  'PlatformColor',
  'Pointer Events',
  'RTLExample',
  'Share',
  'Timers',
  'AppTheme',
  'WebSocket',
  'Transforms',
  //  '<LegacyControlStyleTest>',
  //  '<LegacyTransformTest>',
  //  '<LegacyTextInputTest>',
  //  '<LegacyLoginTest>',
  //  '<LegacyDirectManipulationTest>',
  //  '<LegacyImageTest>',
  //  '<LegacyAccessibilityTest>',
];

```

This will be a pain to maintain over time.

This bug is tracking improving the test infrastructure so these pages can be discovered in a more data-driven way.  There isn't a straightforward solution as Appium xpath selectors seem to be the way to do this but xpath is known to be brittle.  Appium has a custom selector feature we might be able to leverage for this.

### PR DESCRIPTION
**Pull request by NickGerleman**
Commits: https://github.com/facebook/react-native/compare/9b973f32e...5bc67b658

This brings the RNTester Redesign into our repo. Most things work relatively out of the box, with our custom examples updated to conform to new naming. The redesign adds the ability to set optionally set category and documentation URL per example, which I didn't do for our custom examples since I'm unsure it will be useful.

We need to change E2ETest logic to correctly click the right tabs for API example vs component example since they are now split. 
The visual changes also seem to effect tree dumps. Will need to regenerate them and copy the new ones in CI.

There are some subtle issues like #6303 where we don't see examples resize correctly on Window resize correctly.

A crash showed up where we don't correctly support nesting non-Text components inside of Text. This is tracked by #6315, but we now handle this more gracefully and yellowbox instead.

We need to continue some forking for react-native-win32 which currently doesn't expose an AsyncStorage implementation in Rex. Filed #6316 to expose AsyncStorage in Rex instead. I also discovered the package/web debugging path for testing Rex was broken after the RNTester extraction, so package scripts and the react-native-win32 metro config are updated for this.

We also need to remove new calls to Dimensions in react-native-win32 which are intentionally NYI, but can likely be removed upstream in favor of styling changes.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6318)

### CODE REVIEW

File: packages/@react-native-windows/tester/src/js/RNTesterApp.windows.js [hunk 1]
```diff
@@ -127,52 +162,70 @@ const RNTesterExampleListViaHook = ({
...
 
 class RNTesterApp extends React.Component<Props, RNTesterNavigationState> {
   _mounted: boolean;
 
+  constructor() {
+    super();
+
+    // RNTester App currently uses Async Storage from react-native for storing navigation state
+    // and bookmark items.
+    // TODO: Add Native Async Storage Module in RNTester
```
Comment (kmelmon) [inline packages/@react-native-windows/tester/src/js/RNTesterApp.windows.js:198]: >// TODO: Add Native Async Storage Module in RNTester [](start = 4, length = 52)

File issue to track this.

Comment (NickGerleman) [reply packages/@react-native-windows/tester/src/js/RNTesterApp.windows.js:198]: This is from upstream

File: packages/@react-native-windows/tester/src/js/examples-win/Flyout/FlyoutExample.windows.tsx [hunk 2]
```diff
@@ -251,7 +251,7 @@ class FlyoutExample extends React.Component<{}, IFlyoutExampleState> {
 }
 
 export const displayName = (_undefined?: string) => {};
-export const title = '<Flyout>';
+export const title = 'Flyout';
```
Comment (kmelmon) [inline packages/@react-native-windows/tester/src/js/examples-win/Flyout/FlyoutExample.windows.tsx:254]: add category?

Comment (NickGerleman) [reply packages/@react-native-windows/tester/src/js/examples-win/Flyout/FlyoutExample.windows.tsx:254]: Touch on it briefly in the description. Without a category, the default is to say "Other". Can change them to "UI" if we want proper categories.

File: packages/E2ETest/wdio/test/VisitAllPages.spec.ts [hunk 3]
```diff
@@ -61,21 +70,26 @@ let pages = [
...
-      console.log('loading page ' + page);
-      HomePage.goToTestPage(page);
-      HomePage.backToHomePage();
-    });
-  });
+  componentExamples.forEach(ex => addTestPage(ex, 'component'));
+  apiExamples.forEach(ex => addTestPage(ex, 'api'));
 });
+
+function addTestPage(name: string, type: 'api' | 'component') {
```
Comment (kmelmon) [inline packages/E2ETest/wdio/test/VisitAllPages.spec.ts:80]: >addTestPage [](start = 9, length = 11)

nit:  consider renaming addTestPage => visitTestPage

**Review by kmelmon — APPROVED** [review]
:shipit:

**Comment by NickGerleman** [comment]
I think this might have introduced extra instability to E2ETest. Need to test a bit further.

**Comment by NickGerleman** [comment]
Instability seemed to be due to perf issues. It was discovered that we were running E2ETest with PageHeap enabled, which was nuking perf. Talked to Keith about disabling it, and it seems to make sense to since it dramatically improves perf/reliability.

File: packages/@react-native-windows/tester/src/js/examples-win/Flyout/FlyoutExample.windows.tsx (hunk 2 above)
Comment (NickGerleman) [reply packages/@react-native-windows/tester/src/js/examples-win/Flyout/FlyoutExample.windows.tsx:254]: Added categories to all of our custom examples

**Pull request merged**